File: pkg/tg/message.go
package tg

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"time"
)

type TgChat struct {
	Id       int64  `json:"id"`
	Type     string `json:"type"`
	Username string `json:"username"`
}

type TgMessage struct {
	Id   int64  `json:"message_id"`
	Date int64  `json:"date"`
	From TgChat `json:"from"`
	Chat TgChat `json:"chat"`
	Text string `json:"text"`
}

type TgResponse[T any] struct {
	Ok     bool   `json:"ok"`
	Result T      `json:"result"`
}

func getBotURL(path ...string) string {
	botUrl := fmt.Sprintf("https://api.telegram.org/bot%s", tgConfig.Get("token"))
	if len(path) > 0 {
		fullUrl, _ := url.JoinPath(botUrl, path...)
		return fullUrl
	} else {
		fullUrl, _ := url.JoinPath(botUrl, "/")
		return fullUrl
	}
}

func SendMessage(
	chatId string,
	message string,
) (TgMessage, error) {
	url := getBotURL("sendMessage")
	body, err := json.Marshal(map[string]interface{}{
		"chat_id": chatId,
		"text":    message,
	})
	if err != nil {
		slog.Error("Cannot serialize message", "message", message, "error", err)
		return TgMessage{}, err
	}

	slog.Debug("Sending message", "message", message, "url", url, "body", string(body))
	bufBody := bytes.NewBuffer(body)
	resp, err := http.Post(url, "application/json", bufBody)
	if err != nil {
		slog.Error("Cannot send message", "message", message, "error", err)
		return TgMessage{}, err
	}
	defer resp.Body.Close()

	slog.Debug("Received response", "status", resp.StatusCode, "message", message)

	var messageResponse TgResponse[TgMessage]
	if err := json.NewDecoder(resp.Body).Decode(&messageResponse); err != nil {
		slog.Error("Cannot decode response", "message", message, "error", err)
		return TgMessage{}, err
	}

	return messageResponse.Result, nil
}

func SendPing(
	chatId string,
	message string,
	transient bool,
	timeout time.Duration,
) error {
	if message == "" {
		message = "Heads up!"
	}

	origMessage, err := SendMessage(chatId, message)
	if err != nil {
		slog.Error("Cannot send ping", "chatID", chatId, "error", err)
		return err
	}

	messageId := origMessage.Id
	updateUrl := getBotURL("getUpdates")

	updateParams := struct { AllowedUpdates []string `json:"allowed_updates"`; Offset int64 `json:"offset"` } {
		Offset: -1,
		AllowedUpdates: []string{ "message", "message_reaction" },
	}

	pollResult := make(chan bool)
	cleanup := []int64 { messageId }
	messageSentTime := time.Unix(origMessage.Date, 0)

	go func() {
		for {
			ok := func() bool {
				jsonUpdateParams, _ := json.Marshal(updateParams)
				bufUpdateParams := bytes.NewBuffer(jsonUpdateParams)
				resp, err := http.Post(updateUrl, "application/json", bufUpdateParams)
				if err != nil {
					slog.Warn("Failed to poll updates", "error", err)
					return false
				}
				defer resp.Body.Close()
	
				var messageResponse TgResponse[[1]map[string]any]
				json.NewDecoder(resp.Body).Decode(&messageResponse)
				
				slog.Debug("Received updates", "response", messageResponse)
				
	
				for _, message := range messageResponse.Result {
					if message["message_reaction"] != nil {
						if updateMessageId := message["message_reaction"].(map[string]any)["message_id"]; updateMessageId != nil && int64(updateMessageId.(float64)) == messageId {
							slog.Debug("Received pong by reaction", "messageId", messageId)
							return true
						}
					} else if message["message"] != nil {
						updateMessageTime := int64(message["message"].(map[string]any)["date"].(float64))
						if updateMessageTime > messageSentTime.Unix() {
							slog.Debug("Received pong by message", "messageId", messageId)

							updateMessageId := int64(message["message"].(map[string]any)["message_id"].(float64))
							cleanup = append(cleanup, updateMessageId)

							return true
						}
					}
	
					if updateId := message["update_id"]; updateId != nil {
						updateParams.Offset = int64(message["update_id"].(float64)) + 1
					}
				}
				return false
			}()
			if ok {
				pollResult <- true
				break
			}
			slog.Debug("Waiting for message update", "messageId", messageId)
			time.Sleep(5 * time.Second)
		}
	}()

	select {
	case <-pollResult:
		now := time.Now()
		elapsed := now.Sub(messageSentTime)
		slog.Debug("Received response", "elapsed_min", elapsed.Minutes())

		if transient {
			cleanupUrl := getBotURL("deleteMessages")
			cleanupBody, _ := json.Marshal(map[string]interface{}{
				"chat_id": chatId,
				"message_ids": cleanup,
			})
			cleanupBufBody := bytes.NewBuffer(cleanupBody)
			if cleanupResp, err := http.Post(cleanupUrl, "application/json", cleanupBufBody); err == nil {
				cleanupResp.Body.Close()
			}
		}

		return nil
	case <-time.After(timeout):
		slog.Warn("Ping message update timed out", "messageId", messageId)
		fmt.Fprintf(os.Stderr, "Timeout after %s\n", timeout)
		os.Exit(1)
		return nil
	}
}

File: src/tg/message.rs
use std::thread;
use std::time::{Duration, Instant};

use anyhow::{Context, Result};
use serde::Deserialize;
use serde_json::{Value, json};

#[derive(Deserialize)]
pub struct TgResponse<T> {
    #[allow(dead_code)]
    pub ok: bool,
    pub result: T,
}

#[derive(Deserialize)]
pub struct TgMessage {
    #[serde(rename = "message_id")]
    pub id: i64,
    pub date: i64,
}

pub struct TgClient {
    token: String,
    pub chat_id: String,
}

impl TgClient {
    pub fn new(token: &str, chat_id: &str) -> Self {
        Self {
            token: token.to_string(),
            chat_id: chat_id.to_string(),
        }
    }

    fn bot_url(&self, path: &str) -> String {
        format!("https://api.telegram.org/bot{}/{}", self.token, path)
    }

    pub fn send_message(&self, text: &str) -> Result<TgMessage> {
        let url = self.bot_url("sendMessage");
        let body = json!({
            "chat_id": self.chat_id,
            "text": text,
        });

        log::debug!("Sending message: {text} to {url}");

        let resp: TgResponse<TgMessage> = ureq::post(&url)
            .send_json(&body)
            .context("Failed to send message")?
            .body_mut()
            .read_json()
            .context("Failed to decode sendMessage response")?;

        Ok(resp.result)
    }

    pub fn send_ping(&self, text: &str, transient: bool, timeout: Duration) -> Result<()> {
        let orig = self.send_message(text)?;
        let message_id = orig.id;
        let message_sent_time = orig.date;

        let mut offset: i64 = -1;
        let mut cleanup = vec![message_id];
        let deadline = Instant::now() + timeout;

        loop {
            if Instant::now() > deadline {
                eprintln!("Timeout after {}s", timeout.as_secs());
                std::process::exit(1);
            }

            let update_body = json!({
                "allowed_updates": ["message", "message_reaction"],
                "offset": offset,
            });

            let updates: Result<TgResponse<Vec<Value>>> = (|| {
                let resp = ureq::post(&self.bot_url("getUpdates"))
                    .send_json(&update_body)
                    .context("Failed to poll updates")?;
                let parsed = resp
                    .into_body()
                    .read_json()
                    .context("Failed to decode getUpdates response")?;
                Ok(parsed)
            })();

            match updates {
                Ok(resp) => {
                    for update in &resp.result {
                        // Check for reaction on our message
                        if let Some(reaction) = update.get("message_reaction")
                            && let Some(mid) = reaction.get("message_id")
                            && mid.as_i64() == Some(message_id)
                        {
                            log::debug!("Received pong by reaction");
                            if transient {
                                self.delete_messages(&cleanup);
                            }
                            return Ok(());
                        }

                        // Check for new message after our sent time
                        if let Some(msg) = update.get("message")
                            && let Some(date) = msg.get("date").and_then(|d| d.as_i64())
                            && date > message_sent_time
                        {
                            log::debug!("Received pong by message");
                            if let Some(mid) = msg.get("message_id").and_then(|m| m.as_i64()) {
                                cleanup.push(mid);
                            }
                            if transient {
                                self.delete_messages(&cleanup);
                            }
                            return Ok(());
                        }

                        // Advance offset
                        if let Some(uid) = update.get("update_id").and_then(|u| u.as_i64()) {
                            offset = uid + 1;
                        }
                    }
                }
                Err(e) => {
                    log::warn!("Failed to poll updates: {e}");
                }
            }

            log::debug!("Waiting for message update for message_id={message_id}");
            thread::sleep(Duration::from_secs(5));
        }
    }

    fn delete_messages(&self, message_ids: &[i64]) {
        let body = json!({
            "chat_id": self.chat_id,
            "message_ids": message_ids,
        });

        if let Err(e) = ureq::post(&self.bot_url("deleteMessages")).send_json(&body) {
            log::warn!("Failed to delete messages: {e}");
        }
    }
}
